### PR TITLE
Document compatibility tests

### PR DESCRIPTION
**Pull request by cocreature**
Since I’ve explained at various points what is being tested and this
is something that is useful to non-devs as well, I’ve written up the
tests we currently have and what exactly is covered.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
cc @bame-da since you might be interested in this.

File: compatibility/README.md [hunk 1]
```diff
@@ -13,3 +13,76 @@ to be there for this workspace to use.
...
+cron job. On each PR we only include HEAD and the latest stable
+release.
+
+#### Cross-version compatibility between ledger-api-test-tool
+
+We test that the `ledger-api-test-tool` of a given version passes
+against Sandbox next of another version. We test all possible version
+combinations here to ensure forwards and backwards compatibility. The
+`ledger-api-test-tool` includes a DAR built using a compiler from
+the same SDK version so this also ensures that we can load a DAR from
```
Comment (garyverhaegen-da) [inline compatibility/README.md:34]:
```suggestion
the same SDK version so this also ensures that sandbox can load a DAR from
```

File: compatibility/README.md [hunk 2]
```diff
@@ -13,3 +13,76 @@ to be there for this workspace to use.
...
+
+2. Iterate over SDK versions in order. For each version:
+
+   1. Start sandbox of that SDK version.
+
+   2. Run a custom Scala binary that interacts with the ledger and
+      records results.
+
+   3. Validates the results. This includes verifying that the
+      transaction streams are the same after the migration and the the
```
Comment (garyverhaegen-da) [inline compatibility/README.md:55]:
```suggestion
      transaction streams are the same after the migration and that the
```

File: compatibility/README.md [hunk 3]
```diff
@@ -13,3 +13,76 @@ to be there for this workspace to use.
...
+combinations here to ensure forwards and backwards compatibility. The
+`ledger-api-test-tool` includes a DAR built using a compiler from
+the same SDK version so this also ensures that we can load a DAR from
+a different SDK version.
+
+Since all our JVM ledger clients use the same client libraries we
+consider the `ledger-api-test-tool` to be a good proxy and if things
+are not covered it should be extended.
+
+#### Data-continuity for Sandbox classic
```
Comment (garyverhaegen-da) [inline compatibility/README.md:41]: Given the current description, it's unclear to me whether this involves creating any contract in a past version and accessing said contracts after migration. The DAR could be pure library.

File: compatibility/README.md [hunk 4]
```diff
@@ -13,3 +13,76 @@ to be there for this workspace to use.
...
+that only iterates through stable versions up to HEAD. This ensures
+that both individual migrations are correct as well as the migrations
+from one stable version to the next work as a whole.
+
+#### Backwards-compatibility for DAML Script
+
+We test that the DAML Script runner from a given SDK version can load
+DARs built against the DAML Script library from an older SDK. We only
+guarantee backwards compatibility here.
+
```
Comment (garyverhaegen-da) [inline compatibility/README.md:68]:
```suggestion
```

File: compatibility/README.md (hunk 3 above)
Comment (cocreature) [reply compatibility/README.md:41]: Good point, I’ve spelled out the “interactions with the ledger” a bit more explicitly. Does that help?

Comment (garyverhaegen-da) [reply compatibility/README.md:41]: 👍

**Pull request merged**